Add media_player.dlna_dmr component

StevenLooman opened the pull request:
## Description:


**Related issue (if applicable):** -

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5474

## Example entry for `configuration.yaml` (if applicable):
```yaml
media_player:
  - platform: dlna_dmr
    url: http://192.168.178.10:49152/description.xml
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54


For discovery, see home-assistant/netdisco#196

StevenLooman committed:
Add media_player.dlna_dmr component

StevenLooman commented:
The version of `netdisco` needs to be updated as well, but this depends on an new release of it.

Extra tasks:

  - [ ] Release new version of `netdisco`
  - [ ] Change `netdisco` version requirement

StevenLooman commented:
Lets add some documentation with regard to some design decisions and implementation details. It might make a review a bit easier, as the module got bigger than expected.

The decorators `requires_action` (line 71) and `requires_state_variable` (line 110) are used later on, to ensure methods like `DlnaDmrDevice.async_set_volume_level()` have the services with actions.

The advantage is that if service/action is available, the action is passed on to the method and can be used. No need to look it up in each method. If it is not available, handling (i.e., raising NotImplemented) is done automatically.
The disadvantage is that it adds more code, likely harder to maintain. The `DLNA` standard requires the `RenderingControl` service to be implemented (thus, these checks could be omitted, in theory), but the `AVTransport` service is optional.
I am unsure if `DlnaDmrDevice.supported_features()` reports not being able to pause the device, if the service cannot be called. If so, we should be able to remove the decorators and make the module little bit smaller/simpler.


The `UpnpNotifyView` is there to receive notifications from the device. For example, if you turn up the volume, the device can report this change.
Notice that some devices play less nice (such as the [upmpdcli](https://www.lesbonscomptes.com/upmpdcli/)) which send updates before the initial 'registration' request (done in `DlnaDmrDevice._async_init_device()` is completed. In other words, the used SID (unique identifier shared between the client/device, generated by the device) is not known when we already receive a callback. As a work-around, a small backlog (see `UpnpNotifyView._backlog`) was added.


Finally, some devices, such as my Samsung TV, is rather picky about the source it wants to play from. Since `play_media` is implemented (`DlnaDmrDevice.async_play_media()`), the TTS service pops up. It is expected that users will want to use this, or at least test it. If the TTS service is used, it calls `play_media` with a parameter containing the URL to the TTS-file. Also, others have already tried to play media served via `Home Assistant`. Without the `PickyDeviceProxyView`, my TV refuses to play it, reporting an error. This view adds some headers and functionality (`Accept-Ranges` and some DLNA specific headers), such that more picky devices can play media as well served by `Home Assistant`.


Feel free to suggest changes. In hindsight, it might be better to remove the checks around method which use he DLNA `RenderingControl` service (volume setting/muting.)

dgomes commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,909 @@
+    add_devices([device])
+
+
+@asyncio.coroutine
```
We have adopted PEP 492, please update accordingly, eg:

```
@asyncio.coroutine
def function():`
```
to 

```
async def function
```

`yield from` becomes `await`

StevenLooman replied:
Will do.

dgomes commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,909 @@
+        return supported_features
+
+    @property
+    @requires_state_variable('RC', 'Volume')
```
Properties must not execute any IO

All property values must be populated in async_update(), if I understand correctly requires_state_variable will trigger some UPnP message exchange

StevenLooman replied:
This does not do any IO. In `DlnaDmrDevice._async_init_device()` the specification of the device is parsed. Via the `NotifyView` the any values for state variables are stored. These methods simply read whats already stored.

dgomes commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,909 @@
+                    PickyDeviceProxyView.DLNA_CONTENT_FEATURES.replace('17',
+                                                                       '00')
+
+        meta_data = """
```
Lets keep this XML string as constants defined at the beginning of the file

dgomes commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,909 @@
+        self.schedule_update_ha_state()
+
+    @property
+    def supported_features(self):
```
Have a look at https://github.com/home-assistant/architecture/issues/1

dgomes commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,909 @@
+            del self._registered_services[sid]
+
+
+class PickyDeviceProxyView(HomeAssistantView):
```
I understand the "PickyDevice" issue, but we should be able to hide this issue from HomeAssistant and move its handling to the supporting library. 

I understand that it is no easy task, but today it's Samsung, tomorrow some 10 other different brands, each picky on something different.

StevenLooman replied:
Fair enough, I'll remove it.

But, people will try to use TTS services and report their devices not playing.
Not sure how to move this to another library. This requires starting a new webserver, mixing it with home assistants event loop, if possible? Is there a way to disable the TTS service for a specific device, from within the component?

dgomes commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,909 @@
+
+    cfg_extra = {
+        CONF_MAX_VOLUME: config.get(CONF_MAX_VOLUME),
+        CONF_PICKY_DEVICE: config.get(CONF_PICKY_DEVICE) or is_picky,
```
config.get(CONF_PICKY_DEVICE, is_picky)

dgomes reviewed (changes requested):
This is indeed a very large PR, but a welcome one :)

I've only gone through the code very fast, but leave you already some items to address.

StevenLooman committed:
Merge remote-tracking branch 'upstream/dev' into dev

StevenLooman committed:
PEP 492

StevenLooman committed:
Move DIDL-template up

StevenLooman committed:
Remove max_volume-override option

StevenLooman committed:
Remove picky_device support

StevenLooman committed:
Use DEFAULT_NAME

StevenLooman committed:
Make supported_features static

StevenLooman commented:
I have changed things according to your feedback.

One question though, about `supported_features`. The volume/mute control are mandatory according to the DLNA standard. But play/pause/stop/next/previous/play_media are optional. Should the device set `supported_features` accordingly, during initialization, such as the `Open Home` component? Or always report the features and error upon usage?

dgomes commented:
supported_features has an impact in the frontend, thats one of the reasons to exposed in initialisation everything it might support in the "future".

In my understanding of the architecture discussion, I would say that optional should be part of supported_features. 

But someone else will necessarily come in and do a 2nd review and can comment on this view :)

dgomes commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -190,7 +190,7 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
     if config.get(CONF_URL) is not None:
         url = config.get(CONF_URL)
-        name = config.get(CONF_NAME)
+        name = config.get(CONF_NAME, DEFAULT_NAME)
```
No need to pass DEFAULT_NAME, voluptuous will have already set it to DEFAULT_NAME

StevenLooman committed:
Remove unneeded argument

StevenLooman committed:
Proper module-docstring

StevenLooman committed:
Add http dependency

StevenLooman commented:
Thanks!


Op ma 11 jun. 2018 17:57 schreef Diogo Gomes <notifications@github.com>:

> *@dgomes* approved this pull request.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/14749#pullrequestreview-127631813>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAsNIzj0ivIFW0T64KuvOg_sA5X93C2dks5t7pNogaJpZM4UXTVc>
> .
>

rytilahti commented:
As a side note, should this be added to handle the DLNA from netdisco (`components/discovery.py`)?
```
unknown service discovered: DLNA {'host': '192.168.XX.XX', 'port': 9999, 'ssdp_description': 'http://192.168.XX.XX:9999/XXXXXXXX/Upnp/device.xml', 'name': '小米互联网XXXX', 'model_name': 'S602', 'model_number': 'S602', 'serial': None, 'manufacturer': 'Mi, Inc.', 'udn': 'uuid:XXXXXX'}
```

rytilahti reviewed (commented):
Some misc comments, I think the protocol level handling should be done in an external library instead of here. Basic functionality tested briefly to work.

StevenLooman commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,693 @@
+
+DEPENDENCIES = ['http']
+
+DEFAULT_NAME = 'DLNA_DMR'
```
DMR stands for 'Digital Media Renderer'. Should we use this?

Also, should the file itself also be renamed?

rytilahti replied:
I suppose that is fine too. Wrt naming of the file, I think it's fine to use abbreviations.

StevenLooman commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,693 @@
+  <upnp:class>{upnp_class}</upnp:class>
+  <res protocolInfo="http-get:*:{mime_type}:{dlna_features}">{media_url}</res>
+</item>
+</DIDL-Lite>"""
```
While I agree, 'the' backend library is specifically for UPnP. DLNA is a standard on top of UPnP. DIDL is another standard, used by DLNA. Unfortunatly, I couldn't find any Python DIDL packages on PyPI. Moving this to the UPnP library would introduce things not related to UPnP itself.

I'll try to come up with a solution.

rytilahti replied:
Is a pure UPnP library useful nowadays without supporting DLNA specific interfaces? I'm living under the impression that it's nowadays pretty much just DLNA (in which case I would include DLNA specifics to the same library).

StevenLooman replied:
Perhaps netdisco should get a review then as well. Currently it targets DLNA, but not 'generic' UPnP devices. Lots of routers talk UPnP as well, for example.

StevenLooman replied:
Thank you for the pointer. I've seen the SoCo-implementation, but it is incomplete (only the parts required for SoCo are there.) I've built python-didl-lite and will be using this.

StevenLooman replied:
@rytilahti What you mean with:
> @StevenLooman looks like the DLNA check uses 'urn:schemas-upnp-org:device:MediaServer:1', which is actually wrong for this PR :-)

rytilahti replied:
Ok, great, I think I may have a use for that library elsewhere :+1:

What I meant is that netdisco checks for MediaServer too: https://github.com/home-assistant/netdisco/blob/master/netdisco/discoverables/DLNA.py - somehow I missed that it has DNR too there, so I think that should be fine anway.

StevenLooman commented:
@rytilahti Thank you for the review. I'll change things accordingly.

rytilahti commented:
You may also want to capture the exceptions raised by the library:
```
Jun 23 15:57:00 nuc hass[28733]: 2018-06-23 15:57:00 ERROR (MainThread) [homeassistant.core] Error executing service <ServiceCall media_player.media_play_pause: entity_id=['media_player.dlna_dmr_2']>
Jun 23 15:57:00 nuc hass[28733]: Traceback (most recent call last):
Jun 23 15:57:00 nuc hass[28733]:   File "/home/hass/home-assistant/homeassistant/core.py", line 1007, in _event_to_service_call
Jun 23 15:57:00 nuc hass[28733]:     await service_handler.func(service_call)
Jun 23 15:57:00 nuc hass[28733]:   File "/home/hass/home-assistant/homeassistant/components/media_player/__init__.py", line 441, in async_service_handler
Jun 23 15:57:00 nuc hass[28733]:     await getattr(player, method['method'])(**params)
Jun 23 15:57:00 nuc hass[28733]:   File "/home/hass/home-assistant/homeassistant/components/media_player/dlna_dmr.py", line 531, in async_media_play
Jun 23 15:57:00 nuc hass[28733]:     await action.async_call(InstanceID=0, Speed='1')
Jun 23 15:57:00 nuc hass[28733]:   File "/home/hass/home-assistant/libs/async_upnp_client/async_upnp_client/async_upnp_client.py", line 411, in async_call
Jun 23 15:57:00 nuc hass[28733]:     raise UpnpError('Error during async_call(), status: %s, body: %s' % (status_code, response_body))
Jun 23 15:57:00 nuc hass[28733]: async_upnp_client.async_upnp_client.UpnpError: Error during async_call(), status: 500, body: <?xml version="1.0"?><s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/"><s:Body><s:Fault><faultcode>s:Client</faultcode><faultstring>UPnPError</faultstring><detail><UPnPError xmlns="urn:schemas-upnp-org:control-1-0"><errorCode>701</errorCode><errorDescription>Action Play failed
```

StevenLooman committed:
Remove additional_configuration options, no longer used

StevenLooman committed:
Change default name to 'DLNA Digital Media Renderer'

StevenLooman commented:
@rytilahti The specific error you mention is when playing media at your Samsung TV, probably when trying to use the TTS server? The pull request included a work-around for devices which are rather picky about the source to be played (e.g., Samsung TVs), but was removed from this pull request.

StevenLooman commented:
I'm going to fix the issues in async_upnp_client and work on the parts from the reviews. Hopefully this will be done soon, but currently days can be busy.

rytilahti commented:
@StevenLooman it may have been the samsung tv indeed, I was just testing out if any of my devices is able to use the TTS at all, with no great results :-)

StevenLooman commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,693 @@
+    @property
+    def unique_id(self) -> str:
+        """Return an unique ID."""
+        return "{}.{}".format(__name__, self._url)
```
This bugged me before and I'm glad you brought it up. The UUID is not known during the initial lifetime of the component. I.e., the component is initialized and on the first update the description file is parsed and the UUID becomes known.

Thus, the component might change unique_ids during its lifetime from None to a value. Is this acceptable for Home Assistant?

When the device is found via discovery, the same holds. Once the description file is parsed, the component knows the UUID.

rytilahti replied:
I'm not sure if it's okay to change the UUID after the initialization, as I think it will just get asked once during/after the setup phase. For non-discovered devices you could setup the device first after its service description is parsed, for discovered devices it is much simpler (as netdisco already parses the file during the discovery, and it could thus pass it onwards if not done already).

StevenLooman committed:
Use python-didl-lite for DIDL-Lite-xml construction/parsing

StevenLooman committed:
Handle NOT_IMPLEMENTED for UPnP state variables RelativeTimePosition and CurrentMediaDuration

StevenLooman committed:
Use UPnP-UDN for unique_id

StevenLooman commented:
@rytilahti How should exceptions from libraries be handled? I.e., what should the component do when an exception is caught?

StevenLooman committed:
Proper handling of upnp events

StevenLooman committed:
Keeping flake8 happy

StevenLooman committed:
Update requirements_all.txt

StevenLooman commented:
Changes:

- Changed according to reviews;
- Moved DIDL-Lite creation/parsing to python-didl-lite;
- Fixes in event handling in async_upnp_client.

How does Home Assistant handle devices which are discovered multiple times? I.e., a device is discovered with IP 10.0.0.100, device restarts, gets a new DHCP lease with IP 10.0.0.101 and Home Assistant discovers it again.

Ideally, we would not bind the URL to a device. The UDN should be enough (used for unique_id now, as per review.) However, as IP change noted above, how will Home Assistant handle this? The unique_id of the 'two' devices will be the same, but the component/implementation stays alive, try to find a (living) device at the old IP?

rytilahti commented:
re: exceptions, the component should handle those without crashing the component/platform and hopefully informing the user through logger about an error.

re: unique id, I have assumed that it will be replaced as it would not make any sense otherwise, but have never tested it myself.

rytilahti commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,717 @@
+        await action.async_call(InstanceID=0,
+                                CurrentURI=media_id,
+                                CurrentURIMetaData=meta_data)
+        await asyncio.sleep(0.25)
```
Why sleep here? Making sure the file gets queued before calling play? Is 0.25 guaranteed to be enough for that?

StevenLooman replied:
My DLNA-enabled devices gave problems when the sleep was not added. I cannot guarantee if 0.25 second will be enough.

Another option might be spin-waiting for a state variable change. Would that be ok? Also, how long is the component allowed to wait?

StevenLooman replied:
With spin-waiting I meant waiting, up to a maximum time, checking if a given state variable has been changed via the notification system.

rytilahti replied:
Yeah, that upnp api sounds really ugly on this regard, but I suppose the sleep (or wait for a metadata notification change) is required to set it into the mode where it can play the new file, and thus it really depends on the device if 0.25 is enough or not.

I'm unsure what's the nicest way to go here, tbh, so I hope someone else will chime in here. As it's asyncio.sleep, it won't be blocking anything else so making the sleep time higher would not be a problem (but it hampers the user experience, of course). I would go to the route of "spin-waiting" for a specific time, as for properly working devices that won't be an issue (+ it will wait no longer than required).

rytilahti commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,717 @@
+
+        # send play command
+        await self.async_media_play()
+        await asyncio.sleep(0.25)
```
Why sleep here?

StevenLooman replied:
Guess a left-over/duplicate of testing. I'll remove it and test it.

rytilahti commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,717 @@
+CONF_UDN = 'udn'
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_URL): cv.string,
+    vol.Required(CONF_UDN): cv.string,
```
Please do not make UDN entry a requirement as it can be read from the service desc file. It is painful enough to find out what URL you should be using, requiring users to parse the XML for UDN is a no-go in my eyes.

StevenLooman replied:
Ok, will change.

rytilahti commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,717 @@
+        hass.data[__name__] = {}
+
+    hass.async_run_job(start_notify_view, hass)
+    notify_view = hass.data[__name__]['notify_view']
```
I'm getting an exception at this point, most likely that async job should be a synchronous one, or this variable should first be accessed after it has signaled to be ready:
```
Jun 30 17:49:02 nuc hass[18016]:   File "/home/hass/home-assistant/homeassistant/components/media_player/dlna_dmr.py", line 204, in setup_platform
Jun 30 17:49:02 nuc hass[18016]:     notify_view = hass.data[__name__]['notify_view']
Jun 30 17:49:02 nuc hass[18016]: KeyError: 'notify_view'
```

rytilahti commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,717 @@
+                                             data=body)
+            response_body = await response.text()
+            await response.release()
+        await asyncio.sleep(0.25)
```
Unnecessary sleep?

StevenLooman replied:
Might be a duplicate of another sleep. I'll remove it and test it.

StevenLooman replied:
Now I remember... I was having problems like these:
ConnectionResetError: Cannot write to closing transport
aiohttp.client_exceptions.ClientOSError: [Errno None] Can not write request body for http://192.168.178.10:49152/ctl/AVTransport

Any clue on how to solve these?

StevenLooman replied:
The actual tracebacks:
```
Traceback (most recent call last):
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/client_reqrep.py", line 474, in write_bytes
    await self.body.write(writer)
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/payload.py", line 181, in write
    await writer.write(self._value)
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/http_writer.py", line 90, in write
    self._write(chunk)
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/http_writer.py", line 57, in _write
    raise ConnectionResetError('Cannot write to closing transport')
ConnectionResetError: Cannot write to closing transport

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/steven/src/ha/home-assistant/homeassistant/core.py", line 1007, in _event_to_service_call
    await service_handler.func(service_call)
  File "/home/steven/src/ha/home-assistant/homeassistant/components/media_player/__init__.py", line 441, in async_service_handler
    await getattr(player, method['method'])(**params)
  File "/home/steven/src/ha/home-assistant/homeassistant/components/media_player/dlna_dmr.py", line 588, in async_play_media
    await self.async_media_play()
  File "/home/steven/src/ha/home-assistant/homeassistant/components/media_player/dlna_dmr.py", line 544, in async_media_play
    await action.async_call(InstanceID=0, Speed='1')
  File "/home/steven/src/ha/async_upnp_client/async_upnp_client/async_upnp_client.py", line 438, in async_call
    await self.service.requester.async_http_request('POST', url, headers, body)
  File "/home/steven/src/ha/home-assistant/homeassistant/components/media_player/dlna_dmr.py", line 325, in async_http_request
    data=body)
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/client.py", line 387, in _request
    await resp.start(conn)
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/client_reqrep.py", line 748, in start
    message, payload = await self._protocol.read()
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/streams.py", line 533, in read
    await self._waiter
aiohttp.client_exceptions.ClientOSError: [Errno None] Can not write request body for http://192.168.178.10:49152/ctl/AVTransport
```

StevenLooman committed:
Make UDN optional

StevenLooman committed:
Merge remote-tracking branch 'upstream/dev' into dev

StevenLooman committed:
Ensure NotifyView is started, before using it

StevenLooman committed:
Only subscribe to services we're interested in

StevenLooman committed:
Don't update state_variables if value has not been changed + minor refactoring

StevenLooman committed:
Improve play_media, follow flow of DLNA more closely

StevenLooman committed:
Hopefully fix ClientOSError problems

StevenLooman committed:
Flake8 fixes

StevenLooman commented:
Please test.

Currently left open:

- [ ] exceptions have to be properly caught

StevenLooman committed:
Keep pylint happy

StevenLooman committed:
Merge remote-tracking branch 'upstream/dev' into dev

StevenLooman committed:
Catch errors and report gracefully

StevenLooman committed:
Update async_upnp_client to 0.11.0

StevenLooman committed:
Don't be so noisy

rytilahti commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,806 @@
+                                             data=body)
+            response_body = await response.text()
+
+        await asyncio.sleep(0.25)
```
I think the problem w/o sleeping here is caused by the transport closing (the old comments have disappeared, so just commenting here). With pure aiohttp I have usually used the context manager which should avoid such cases, but I don't know how this should work with the homeassistant helper.. @balloob do you have an idea here?

rytilahti commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,806 @@
+
+    url = '/api/dlna_dmr.notify'
+    name = 'api:dlna_dmr:notify'
+    requires_auth = False
```
Is there something to worry about if a malicious actor accesses this interface (w/o auth)?

dgomes replied:
this must be blocked to local_network only

StevenLooman replied:
How?

MartinHjelmare replied:
Was this resolved?

rytilahti commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,806 @@
+    async def async_notify(self, request):
+        """Callback method for NOTIFY requests."""
+        if 'SID' not in request.headers:
+            return aiohttp.web.Response(status=412)
```
I think it'd be nice to define non-standard (or more like less commonly used..) response codes as constants. Personally I couldn't say what are 202 nor 412 without looking those up.

StevenLooman replied:
The 412 comes from the UPnP spec. I'll turn them into constants and use those.

StevenLooman replied:
I see some constants in homeassistant/const.py, I'll add the ones I use there and use those.

rytilahti replied:
Better add these to this component only (or the upstream lib, and import them), as they are probably not very interesting for other homeassistant components.

rytilahti commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,806 @@
+        avt_service = action.service
+        state_var = avt_service.state_variable('CurrentTransportActions')
+        if 'Play' not in state_var.value:
+            _LOGGER.debug('Cannot do Play')
```
Crash on sonos:
```
Jul 04 00:28:56 nuc hass[28617]:   File "/home/hass/home-assistant/homeassistant/components/media_player/dlna_dmr.py", line 568, in async_media_play
Jul 04 00:28:56 nuc hass[28617]:     if 'Play' not in state_var.value:
Jul 04 00:28:56 nuc hass[28617]: TypeError: argument of type 'NoneType' is not iterable
```

StevenLooman replied:
Will fix.

rytilahti commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,806 @@
+    @property
+    def supported_features(self):
+        """Flag media player features that are supported."""
+        return SUPPORT_DLNA_DMR
```
Is it possible to induce what features are supported based on api calls? If yes, I think that information should be gathered before the device entity is initialized to avoid "broken"/missing functionality.

StevenLooman replied:
This implies not being able to initialize a device without it being on. I.e., during a (re)start of hass the device (e.g., TV) must be on, otherwise it cannot be initialized when manually configured. Discovery is another story of course, then the device is on/accessible.

`media_player.mpd` had this behavior before and was fairly annoying. It was fixed by #8655. Made at least @pezinek happy. :)

`supported_features` was dynamic before, but was made static due to architectural requirements.

rytilahti replied:
> This implies not being able to initialize a device without it being on. I.e., during a (re)start of hass the device (e.g., TV) must be on, otherwise it cannot be initialized when manually configured.

The platform can raise `PlatformNotReady` exception and homeassistant will keep trying to initialize it later (uses a backoff timer for that), which is to my understanding the correct way to handle that.

rytilahti reviewed (commented):
I think I have reviewed all the parts I can, and tested with a few devices to see that it has at least basic functionality, but I hope someone else could also take a look & test this out before merging. Especially considering that there exists some odd implementations of "almost-dlna" devices.

I suppose I'll do some more testing in the upcoming days to see what functions, but some experiences:

* Sonos play:1 supports only showing playing/paused/idle, and playing & pausing even when previous & next actions are visible. No metadata available (sonos is not dlna certified, so maybe they simply do not do that in a standardized manner).
```
media_duration: 0
media_position: 5
media_position_updated_at: 2018-07-03T22:07:04.929077+00:00
friendly_name: 192.168.XXX.XXX - Sonos PLAY:1
supported_features: 21053
```

* Xiaomi Wifi speaker: play/pause, forward/back, volume control.
```
volume_level: 0.45
media_duration: 245
media_position: 5
media_position_updated_at: 2018-07-03T22:17:08.055190+00:00
media_title: Nena - Irgendwie Irgendwo Irgendwann (Original 1984)
friendly_name: 小米互联网音箱(XXX)
supported_features: 21053
```

* Sony HT-XT3: I suppose this is my only real "DLNA" device. idle when the device is off, but that's probably expected. volume adjustment works, muting works (but state is not propagated to homeassistant, old state is also retained for volume too even when changing works). tts seemed to at least change the media. playing from usb does not seem to update the metadata.
```
volume_level: 0.3
is_volume_muted: false
media_duration: 0
media_title: Home Assistant
friendly_name: HT-XT3
supported_features: 21053
```

Room for improvements:
* Considering you are already listening on events, these could be used for non-polling updates at some point?
* Supported features should be adjusted before initializing the entity to avoid offering missing/broken functionality, or should every "DLNA-certififed" device has these?

dgomes commented:
Have you considered releasing your component as a custom_component on the forum in order to gather beta testers ?

Could be really helpful to gather more experiences to help out removing all the bugs.

StevenLooman commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,806 @@
+    # ensure view has been started
+    hass.async_run_job(start_notify_view, hass)
+    while 'notify_view' not in hass.data[__name__]:
+        time.sleep(0.1)
```
If you have two devices, and both are initialized from another syncworker, then at times I got 2 NotifyView instances. Running it using `hass.async_run_job()` made it ran (for me) ran from the MainThread and thus was able to act as some sort of mutex/lock.

StevenLooman commented:
@dgomes 
> Have you considered releasing your component as a custom_component on the forum in order to gather beta testers ?
> Could be really helpful to gather more experiences to help out removing all the bugs.

Already done, see [this repo](https://github.com/StevenLooman/home-assistant-dlna-dmr/). Now I am asking people who have left issues or contacted me directly to report their findings in this pull request. 
E.g., @yottatsa is reporting here because he contacted me before and I asked him to report here. Thank you for the advice though!

StevenLooman commented:
@yottatsa Thank you for reporting this. I'll take a look at it (and all the other comments) tomorrow.

yottatsa commented:
I found so far that my TV i's replying with `412 Precondition Failed` on `'SUBSCRIBE', self.event_sub_url`, when my hass `base_url` is set to public one. I think that if subscription failed, polling should be working, and it's not.

Also, it seems that URL for SetAVTransportURI should be double-escaped, otherwise, `dlna_handle_notify_last_change` fails to parse `state_var.value`. I didn't found any docs, except https://github.com/UniversalMediaServer/UniversalMediaServer/issues/758. Also, my TV isn't playing this.

Smaller: found unhandled `NOT_IMPLEMENTED` in `media_title` and `media_image_url`, https://gist.github.com/yottatsa/3deb43a07b970395cf9a3e540e00f414.

StevenLooman committed:
Define/use constants for HTTP status codes

StevenLooman committed:
Add discovery entry for dlna_dmr

StevenLooman committed:
More robustness with regard to state variable not being set (yet)

StevenLooman committed:
Keep privates hidden

StevenLooman committed:
Handle NOT_IMPLEMENTED for CurrentTrackMetaData state variable

StevenLooman committed:
Fixes in async_upnp_client + renew UPnP subscriptions regularly

StevenLooman commented:
@yottatsa Can you provide an example with regard to the double escaping?

StevenLooman commented:
Also, if you keep having SUBSCRIBE problems with this new version, please add this to your configuration.yaml to enable traffic dumping via the logger (or dump traffic yourself), and please report requests + responses:
```
logger:
  default: warning
  logs:
    homeassistant.components.media_player.dlna_dmr: debug
    async_upnp_client.async_upnp_client: debug
    async_upnp_client.async_upnp_client.traffic: info
```

StevenLooman committed:
Not too eager

StevenLooman commented:
Same for @rytilahti, the subscription implementation in async_upnp_client has improved. Perhaps your Sony HT-XT3 was not accepting the subscription before due to an invalid header.

The downside for DLNA is that the specification (i.e., [RenderingControl:1](http://upnp.org/specs/av/UPnP-av-RenderingControl-v1-Service.pdf) hardly requires any functionality, but defines the API if the vendor provides the functionality. See table 3, page 19/20: Almost all actions are optional. This is the reason why the methods calling Upnp Actions are decorated.

This was also the reason why `supported_features` was dynamic before. This work as expected in the front-end. E.g., when Previous was not available (provided through `supported_features`), the UI would not show it.

I had not seen [RenderingControl:3](http://www.upnp.org/specs/av/UPnP-av-RenderingControl-v3-Service-20101231.pdf) and [AVTransport:2](http://www.upnp.org/specs/av/UPnP-av-AVTransport-v2-Service-20060531.pdf) before. My assumption was that only RC:1 and AV:1 were available. I'll see what I can do here.

You are right that many devices might not comply (fully) to the spec and thus might not properly work. I think that with some iterating, most devices can be properly supported.
I think this situation is (going to be) more common. I don't know anything about HomeKit, but some vendors will probably (try to) implement it in their devices while not fully complying to spec.

rytilahti commented:
> Same for @rytilahti, the subscription implementation in async_upnp_client has improved. Perhaps your Sony HT-XT3 was not accepting the subscription before due to an invalid header.

Does not seem to be working with the Sony, but I'm not sure why. For transport actions it seems to report only `Stop`.

Great! `GetPositionInfo` seems to work on Sonos now too, although only parts of it are being parsed (`TrackMetaData` contains interesting info, but is it a standard?):
```
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
<s:Body>
<u:GetPositionInfoResponse xmlns:u="urn:schemas-upnp-org:service:AVTransport:1"><Track>3</Track>
<TrackDuration>0:03:30</TrackDuration>
<TrackMetaData>&lt;DIDL-Lite xmlns:dc=&quot;http://purl.org/dc/elements/1.1/&quot; xmlns:upnp=&quot;urn:schemas-upnp-org:metadata-1-0/upnp/&quot; xmlns:r=&quot;urn:schemas-rinconnetworks-com:metadata-1-0/&quot; xmlns=&quot;urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/&quot;&gt;&lt;item id=&quot;-1&quot; parentID=&quot;-1&quot; restricted=&quot;true&quot;&gt;&lt;res protocolInfo=&quot;sonos.com-spotify:*:audio/x-spotify:*&quot; duration=&quot;0:03:30&quot;&gt;x-sonos-spotify:spotify%3atrack%3a1qPea3SGgcs2ODFweqK7hP?sid=9&amp;amp;flags=8224&amp;amp;sn=5&lt;/res&gt;&lt;r:streamContent&gt;&lt;/r:streamContent&gt;&lt;upnp:albumArtURI&gt;/getaa?s=1&amp;amp;u=x-sonos-spotify%3aspotify%253atrack%253a1qPea3SGgcs2ODFweqK7hP%3fsid%3d9%26flags%3d8224%26sn%3d5&lt;/upnp:albumArtURI&gt;&lt;dc:title&gt;Altes Fieber&lt;/dc:title&gt;&lt;upnp:class&gt;object.item.audioItem.musicTrack&lt;/upnp:class&gt;&lt;dc:creator&gt;Die Toten Hosen&lt;/dc:creator&gt;&lt;upnp:album&gt;&amp;quot;Ballast der Republik&amp;quot; plus Jubiläums-Album &amp;quot;Die Geister, die wir riefen&amp;quot;&lt;/upnp:album&gt;&lt;/item&gt;&lt;/DIDL-Lite&gt;</TrackMetaData>
<TrackURI>x-sonos-spotify:spotify%3atrack%3a1qPea3SGgcs2ODFweqK7hP?sid=9&amp;flags=8224&amp;sn=5</TrackURI>
<RelTime>0:02:55</RelTime>
<AbsTime>NOT_IMPLEMENTED</AbsTime>
<RelCount>2147483647</RelCount>
<AbsCount>2147483647</AbsCount>
</u:GetPositionInfoResponse></s:Body></s:Envelope>
```

> The downside for DLNA is that the specification (i.e., RenderingControl:1) hardly requires any functionality, but defines the API if the vendor provides the functionality. See table 3, page 19/20: Almost all actions are optional. This is the reason why the methods calling Upnp Actions are decorated.

I suppose those devices which do not supported the API will not advertise them in their service descriptions? What's the nicest way to test these without running homeassistant, e.g., is it easy to make a console tool to testing the lib against different setups?

I did some more testing, the way you are checking for available actions is broken (as soon as the action string contains whitespaces:
```
available actions: Set, Stop, Pause, Play, X_DLNA_SeekTime, Previous, X_DLNA_SeekTrackNr Set, Stop, Pause, Play, X_DLNA_SeekTime, Previous, X_DLNA_SeekTrackNr
```
and then you check for `if 'Pause' not in actions` (which fails as there is only ` Pause` there). After stripping the whitespaces I'm able to control sonos again.

Neither volume nor mute state is read correctly from that Sonos, volume being 100 always.

As a side note, it looks like every state variable is queried twice for each polling, maybe this is related to the problem you mentioned about double `NotifyView` instances, indicating a problem somewhere deeper?

yottatsa commented:
For testing escaping I've appended HASS TTS URL with senseless parameters: `/api/tts_proxy/...-_google.mp3?hello=es&ca=ping`, and I've checked that it works in a browser. I sent this url using Service call UI. Then I tested it

* with single-escaped url (done by vanilla async-upnp-client code) the TV is stuck in TRANSITIONING (and actually requires power cycle) https://gist.github.com/yottatsa/e3e2f4ca367f8513e4a396ac6bd256da
* with urlencoded query_string everything is working https://gist.github.com/yottatsa/22ad44174e36b7eface9a05e747cd57e:

```
@@ -633,6 +634,9 @@ class DlnaDmrDevice(MediaPlayerDevice):
         """Play a piece of media."""
         # pylint: disable=arguments-differ,unused-variable
         # queue media
+
+        p = urllib.parse.urlparse(media_id)
+        media_id = urllib.parse.urlunparse([p.scheme, p.netloc, p.path, None, urllib.parse.quote_plus(p.query), None])
         meta_data = await self._construct_play_media_metadata(media_type,
                                                               media_id)
         try:

```

I'm not sure if it should be caller side, generic implementation, or dlna_dmr.

StevenLooman committed:
Refactor duplicate code to _current_transport_actions and improve parsing of actions

StevenLooman committed:
Support RC:1 to RC:3 and AVT:1 to AVT:3

StevenLooman commented:
@yottatsa Thank you for the pull request, it is merged. Also, support for RC:1 to RC:3 and AVT:1 to AVT:3 is added.

@rytilahti Thank you for the feedback. The parsing of CurrentTransportActions has been improved, thank you for reporting this.

I'll try to move more DLNA specifics to an external library (maybe to async_upnp_client.) I understand the advantage with regard to debugging and less complexity in Home Assistant. I do think the library will most likely just become a reflection of the Home Assistant media_player API.

Hopefully I have some thing this next week.

rytilahti commented:
> I'll try to move more DLNA specifics to an external library (maybe to async_upnp_client.) I understand the advantage with regard to debugging and less complexity in Home Assistant. I do think the library will most likely just become a reflection of the Home Assistant media_player API.

Great! Personally I'm a great fan of click for creating basic interfaces for my libs, maybe that's something you want to look into. Considering that some of the actions are more than simply "execute X" and require handling notifications, maybe https://github.com/click-contrib/click-repl could be interesting for you there.

Even if it is a reflection of the media_player API, it can still be useful for other, e.g. some people could simple do some simple requests (such as metadata requests) with ease. Personally I would have a couple of use-cases: 1) listening for device notifications, prolly based on async_upnp_client, to remove upnp pieces out from https://github.com/rytilahti/homeassistant-binarysensor-upnp to allow submission to homeassistant), and 2) some devices such as that xiaomi speaker does use its own upnp interfaces for settings management, alarms etc. which could then implemented nicer than my current ad hoc solutions.

westerncanada commented:
Hi Steven, 

I'm happy to confirm that the GGMM E5 Wireless Speakers work very well.

Mark

StevenLooman committed:
Merge remote-tracking branch 'upstream/dev' into dev

StevenLooman committed:
Moved DLNA-specifics to async_upnp_client.dlna.DmrDevice

StevenLooman commented:
I've moved the DLNA-specifics to async_upnp_client. Please do test if things still work.

URL escaping, as mentioned by @yottatsa has been added. @yottatsa, thank you providing a solution.

@rytilahti A upnp-client has been added to async_upnp_client. No direct DLNA support (except for decoding of DLNA events), but you can call actions on services of UPnP devices and subscribe for events. See the readme for additional instructions.

StevenLooman commented:
About the 412 responses, I noticed that a https-URL was sent to the device. Possibly, it doesn't like it. As mentioned before, the internal IP address of the device Home Assistant is running on should be used.

Does Home Assistant provide utilities to provide its internal IP address? I couldn't find anything.

rytilahti commented:
Nice, looks lot cleaner now, I'll do a proper review at some later point, but wanted to quickly mention that `async_upnp_client.dlna` module is currently missing from git.

StevenLooman commented:
Thank you for reporting. I've added it to the repo.

StevenLooman commented:
I think Github lost a comment I posted yesterday. The emulated_hue fires up
its own HTTP server (the HomeAssistant way), I will update the component to
do thi sas well

StevenLooman committed:
Use our own HTTP server to listen for events.

StevenLooman committed:
More clear and explicit log message for easier troubleshooting

StevenLooman commented:
I think I'm done. Please review.

Last commit was to more explicitly show the user on which IP/port is being listened, such that trouble shooting will be easier.

yottatsa commented:
Working on 0.73, failing on 0.75. https://gist.github.com/yottatsa/39a5fc57dd34356d946cbaaa1f4b127e

StevenLooman committed:
Merge remote-tracking branch 'upstream/dev' into dev

StevenLooman committed:
Follow changes by hass, fixes traceback

StevenLooman committed:
Fix not being able to do next

StevenLooman commented:
@yottatsa Thank you for reporting. This is a breaking change by Home Assistant. The component is updated.

I noticed a bug with the next-action. There is a check if the device is able to do Next. The check was missing a not-operator. Fixed it.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+
+DOMAIN = 'dlna_dmr'
```
The domain of this platform is media_player. Remove this constant. Instead define a constant to use when storing data in hass.data. Call i something with the platform name and end with `_DATA`.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+        return hass_data['notify_view']
+
+    # fire up a HTTP server
+    server = HomeAssistantHTTP(
```
Why do we need to start a new server?

Don't use the `HomeAssistantHTTP` class but use the base aiohttp web server class.

See https://github.com/home-assistant/home-assistant/pull/15530.

yottatsa replied:
I've found that if hass is configured to use SSL and/or external FQDN, e.g. `base_url: https://home.yottatsa.name`, then subscription are being refused on some media players. See [comment](https://github.com/home-assistant/home-assistant/pull/14749/#issuecomment-403295181)

StevenLooman replied:
To listen for notifications, as emulated_hue does.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+SETUP_LOCK = asyncio.Lock()
```
Probably store the lock in hass.data.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+        EVENT_HOMEASSISTANT_STOP, server.stop)
+    hass_data['notify_server'] = server
+
+    base_url = 'http://%s:%s/' % (server_host, server_port)
```
Use new style string formatting.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+        return
+
+    if config.get(CONF_URL) is not None:
+        url = config.get(CONF_URL)
```
Don't use `dict.get`when we know that the key is in the dict.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+    """Build a derived instance of UpnpRequester, specific for hass."""
+    from async_upnp_client import UpnpRequester
+
+    class HassUpnpRequester(UpnpRequester):
```
Why do *we* need to define this class?

The requester instance is only used by the interface library. I would expect that library to define this class. The class needs to except the aiohttp session from hass and we just instantiate the class in this module.

StevenLooman replied:
To provide de-coupling from libraries. Now, another developer using async_upnp_client can choose to use requests instead of aiohttp, or another future library.

But I'll extend async_upnp_client a bit to provide it as a utility.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+class DlnaDmrDevice(MediaPlayerDevice):
+    """Representation of a DLNA DMR device."""
+
+    def __init__(self, hass: HomeAssistant, dmr_device, name=None):
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+
+    def __init__(self, hass: HomeAssistant, dmr_device, name=None):
+        """Initializer."""
+        self.hass = hass
```
Remove this.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+        self._name = name
+
+        self._device = dmr_device
+        self._device.on_event = self._on_event
```
Move this callback registration to `async_added_to_hass`. The callback will error if called before the entity has been added to home assistant.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+        except (asyncio.TimeoutError, aiohttp.ClientError):
+            self._available = False
+            _LOGGER.debug("Device unavailable")
+            raise
```
Don't raise. Just log an error and return. Replace the debug log with the error log.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+            except (asyncio.TimeoutError, aiohttp.ClientError):
+                self._available = False
+                _LOGGER.debug("Could not (re)subscribe")
+                raise
```
See above.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+    @property
+    def unique_id(self) -> str:
+        """Return an unique ID."""
+        return "{}.{}".format(__name__, self.udn)
```
Remove `__name__` from the unique id. The core will keep track of platform and domain together with the unique id.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+        """Return an unique ID."""
+        return "{}.{}".format(__name__, self.udn)
+
+    def __str__(self) -> str:
```
Remove this.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+
+    def __repr__(self) -> str:
+        """Repr."""
+        return "<DlnaDmrDevice('{}')>".format(self.udn)
```
The base entity class already defines a repr. Why do we need to overwrite it?

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,542 @@
+        """Set volume level, range 0..1."""
+        try:
+            await self._device.async_set_volume_level(volume)
+        except (asyncio.TimeoutError, aiohttp.ClientError):
```
Almost every entity method contains a similar try... except. Is it possible to break it out into helper or decorator somehow to not repeat it that often?

MartinHjelmare reviewed (commented):
I'm not so familiar with our http servers so another reviewer with that expertise needs to make a second review at least.

See below for my comments and questions.

StevenLooman committed:
Changes after review by @MartinHjelmare

StevenLooman committed:
Linting

StevenLooman commented:
Interesting, my own run didn't find these. Now fixed.

MartinHjelmare commented:
I want a member that's familiar with aiohttp servers to review.

StevenLooman committed:
Merge remote-tracking branch 'upstream/dev' into dev

StevenLooman committed:
Use homeassistant.util.get_local_ip

StevenLooman committed:
Moved upnp event handling to async_upnp_client

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,438 @@
+                              self._async_on_hass_stop)
+
+    @property
+    def udn(self):
```
This seems not needed. We can use `self._device.udn` directly in `unique_id` property.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,438 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+ASYNC_LOCK = asyncio.Lock()
```
Store the lock in hass.data.

MartinHjelmare commented on homeassistant/components/media_player/dlna_dmr.py:
```diff
@@ -0,0 +1,438 @@
+    'async-upnp-client==0.12.2',
+]
+
+DEPENDENCIES = []
```
Remove this.

MartinHjelmare reviewed (changes requested):
If we fix the remaining things I think we can merge.

StevenLooman committed:
Keeping pylint happy

StevenLooman committed:
Changes after review by @MartinHjelmare

MartinHjelmare reviewed (approved):
Great! Can be merged when build passes.

StevenLooman commented:
Thanks! This should do it...

rytilahti commented:
:tada: nice work @StevenLooman and congrats for the merge! :-) I'll give it a try when I have some free time next week and will file bugs if necessary.
